README.md
# ACORNARCHIMEDES-WINDOWS-Arculator
An Archimedes emulator by Tom Walker. 

## License ##

## Author ##
Tom Walker

## Links ##
http://b-em.bbcmicro.com/arculator/index.html


CHANGES.TXT
Version 0.9 (06/02/07) :
- Fixed ARM bugs - Bug Hunter now works, Phaethon no longer crashes on level 3 
- Fixed bug in 82c711 - Nebulus now works on new FDC 
- Fixed timing in modes 48/49 
- Various optimisations 
- Can now emulate ARM3 at 33mhz 
- Now emulates two hard discs 
- Removed software scaling options - improvements to hardware scaling mean these are not 
  necessary anymore 
- Improved FDI support in new FDC - more games supported (eg Wolf 3d, SF3000 etc) with hard 
  disc install 
- Mouse & keyboard shouldn't lock up in some stuff anymore 
- Some other fixes

Version 0.8 (23/10/06) :
- Improved timing. Each model now performs much closer to the original machines (this means
   slower than previous versions).
- Can now emulate MEMC page sizes other than 32k, so all memory configurations from 512k to
   16 megs are now possible.
- Preliminary FDI support in 82c711 FDC.
- Better FDI support in WD1772 FDC, including FDI v2.1 support.
- Fast disc emulation in 82c711 FDC.
- Bugfixes to video code. Man United Europe now looks better. Less flicker in Populous, Zarch
  etc.
- Keyboard/mouse fix, now doesn't freeze in Lemmings 2 and some other games.
- Fixed broken sound in Sensible Soccer and Xenon 2.
- Sound now in stereo.
- Some minor optimisations.
- Seperate ROM + CMOS files for Arthur, RISC OS 2, RISC OS 3 (old FDC) and RISC OS 3 (new FDC).
- Added support for two MAME romsets that run on Archimedes hardware.
- Tracked WD1772 hanging bug down to an incorrectly set CMOS bit! Now fixed.

Version 0.7 (09/05/06) :
Improved video code :
- Display centered always
- Multisync mode support (eg 640x480, 640x512 and 800x600)
- Borders always drawn correctly (eg Zelanites)
- Multisync modes are a useful workaround the desktop hanging bug!
Fixed bug with LDMxx!, !NewLook now works, Arthur desktop now works
Fixed bug in WD1772 FDC, disc writing now works again
Some improvements to ArculFS

version 0.6 (28/04/06) :
Improved FDI support - many more games work
Improved video routines - new hardware scaling option, faster! Also more video
  modes implemented
ArculFS - read files from the Windows file system (read-only at the moment)
Can emulate ARM250

version 0.5 (23/01/06) :
Working 82c711 emulation. This allows high density disc access and avoids the
  hanging bug in 1772 emulation.
Along with 82c711 emulation comes IDE hard disc emulation.
Preliminary FDI support (1772 only).
Increased frequency of keyboard/mouse updates - mouse now much smoother
Slight optimisations to video

version 0.4 (12/10/05) :
Delayed IRQ by one instruction - Lemmings 2 now has sound
Fast disc access
Can disable scanline effect

version 0.35 (05/10/05) :
Fixed reading of R15 in some instructions - Nebulus now works
Sound improved a bit
Fixed RSC instruction (don't think it affects anything though)
Altered ARM timing
Beginnings of new-style Arc emulation (HD floppies + IDE hard discs in theory, not 
  working yet though)
Palette splits slightly more stable (Lotus 2)

version 0.3 (22/08/05) :
Much improved sound (filter still a bit off though)
Slight ARM optimisations - 10% speedup
Bug fixes to VIDC and IOC - TCD Powerscroll and ARMageddon Megademo 3 now work
Emulated prefetch - Elite now works (stupid protection)
Fixed stupid video mode 0 bug
Replaced mouse hack with mouse capture technique
Reduced CPU usage on NT/2000/XP

version 0.25 (22/07/05) :
Disc emulation improved - can now access 720k DOS disc images, write to discs, and up to 4
  drives now supported
Arculator will now try to load a 2 meg file called 'rom.rom' before trying ic24-ic27.rom. 
  This allows use of one-file ROM dumps, and Arthur, RiscOS 2 and ARM Linux ROM dumps (none
  of which really work, but never mind)
Slight changes to video emulation, Zelanites demo now looks okay
Slight ARM optimisations

version 0.2 (28/06/05) :
Added support for dual/quad MEMCs (8 and 16 megs of RAM)
ARM3 emulation
ARM & video optimisations - 25% speed boost
Rewritten video emulation - support midscreen palette/mode changes, correct
  cursor in 256 colour modes, better scrolling in hardware scrolling games, etc.
Reduced sound lag and improved sound overall
Added fullscreen mode.

version 0.1 (24/05/05) :
Couple of CPU bugs fixed - Disc access now okay (hack no longer required), 
many more programs/games run.
Hack to work around other CPU bug.
Better video emulation - more modes, scrolling in James Pond and Chuck
Rock, correct scrolling in supervisor.
New attempt at sound (quite poor).
Better mouse emulation.

version 0.09 (27/01/05) :
Couple of CPU bugs fixed - BASIC and FPU emulation should be fine now.
Better cursor emulation.
Hack to get floppy disc to "work".

pre-release 11 (01/01/04) :
Totally rewritten. RiscOS now goes to desktop + is usable.
Mouse and keyboard also better.

pre-release 10 (30/11/02) :
Offset all CMOS accesses by 64 bytes, now seems to work.
Started floppy emulation (very bugged).
Couple of optimisations.

pre-release 9 (07/08/02) :
More ARM bugs fixed - RiscOS now boots (but with some errors)
Proper VIDC/MEMC/IOC emulation
Keyboard and mouse now emulated (mouse a bit screwy though)
DMA addresses are actually correct now.

pre-release 8 (30/04/02) :
Some more ARM bugs fixed - many more demos/games work.
More SWIs trapped.
Various address modes for video emulation.
Monitor improved.
Video mode changed to 400x300 to allow overscan screens to display.
First binary release.

pre-release 7 (18/04/02) :
Monitor/debugger/disassembler added.
Some stupid ARM bugs fixed (such as LSR not working correctly) - RiscOS gets
further.
SWI faking improved - more demos run.
Alternate address for video emulation (for TCD demo 1)
No longer compiles without DJGPP/Allegro
Correct default palette

pre-release 6 (15/03/02) :
Can now load absolute files. Some demos run.
SWI faking added (only for absolute files).
Couple of ARM bug fixes.
Primitive video emulation (mode 13 only).

pre-release 5 (02/03/02) :
Full 32k MEMC emulation.

SRC/IDE.C
/*Arculator 0.8 by Tom Walker
  IDE emulation*/
#include <stdio.h>
#include "arc.h"

int dumpedread=0;
struct
{
        unsigned char atastat;
        unsigned char error,status;
        int secount,sector,cylinder,head,drive,cylprecomp;
        unsigned char command;
        unsigned char fdisk;
        int pos;
        int spt[2],hpc[2];
} ide;

int idereset=0;
unsigned short idebuffer[256];
unsigned char *idebufferb;
FILE *hdfile[2]={NULL,NULL};
void closeide()
{
        if (hdfile[0]) fclose(hdfile[0]);
        if (hdfile[1]) fclose(hdfile[1]);
}

int skip512[2]={0,0};
void resetide()
{
        int c;
        ide.drive=0;
        ide.atastat=0x40;
        idecallback=0;
        if (!hdfile[0])
        {
                hdfile[0]=fopen("hd4.hdf","rb+");
                if (!hdfile[0])
                {
                        hdfile[0]=fopen("hd4.hdf","wb");
                        putc(0,hdfile[0]);
                        fclose(hdfile[0]);
                        hdfile[0]=fopen("hd4.hdf","rb+");
                }
                atexit(closeide);
        }
        if (!hdfile[1])
        {
                hdfile[1]=fopen("hd5.hdf","rb+");
                if (!hdfile[1])
                {
                        hdfile[1]=fopen("hd5.hdf","wb");
                        putc(0,hdfile[1]);
                        fclose(hdfile[1]);
                        hdfile[1]=fopen("hd5.hdf","rb+");
                }
        }
        idebufferb=(unsigned char *)idebuffer;
        for (c=0;c<2;c++)
        {
                fseek(hdfile[c],0xFC1,SEEK_SET);
                ide.spt[c]=getc(hdfile[c]);
                ide.hpc[c]=getc(hdfile[c]);
                skip512[c]=1;
                if (!ide.spt[c] || !ide.hpc[c])
                {
                        fseek(hdfile[c],0xDC1,SEEK_SET);
                        ide.spt[c]=getc(hdfile[c]);
                        ide.hpc[c]=getc(hdfile[c]);
                        skip512[c]=0;
                        if (!ide.spt[c] || !ide.hpc[c])
                        {
                                ide.spt[c]=63;
                                ide.hpc[c]=16;
                                skip512[c]=1;
                        }
                }
        rpclog("Drive %i - %i %i\n",c,ide.spt[c],ide.hpc[c]);
        }
//        ide.spt=63;
//        ide.hpc=16;
//        ide.spt=16;
//        ide.hpc=14;
}

void writeidew(unsigned short val)
{
//        if (ide.sector==7) rpclog("Write data %08X %04X\n",ide.pos,val);
        idebuffer[ide.pos>>1]=val;
        ide.pos+=2;
        if (ide.pos>=512)
        {
                ide.pos=0;
                ide.atastat=0x80;
                idecallback=200;
        }
}

void writeide(unsigned short addr, unsigned char val)
{
//        if (addr!=0x1F0) rpclog("Write IDE %08X %02X %08X %08X\n",addr,val,PC-8,armregs[12]);
        switch (addr)
        {
                case 0x1F0:
                idebufferb[ide.pos++]=val;
                if (ide.pos>=512)
                {
                        ide.pos=0;
                        ide.atastat=0x80;
                        idecallback=1000;
                }
                return;
                case 0x1F1:
                ide.cylprecomp=val;
                return;
                case 0x1F2:
                ide.secount=val;
                if (!val) ide.secount=256;
//                rpclog("secount %i %02X\n",ide.secount,val);
                return;
                case 0x1F3:
                ide.sector=val;
                return;
                case 0x1F4:
                ide.cylinder=(ide.cylinder&0xFF00)|val;
                return;
                case 0x1F5:
                ide.cylinder=(ide.cylinder&0xFF)|(val<<8);
                return;
                case 0x1F6:
                ide.head=val&0xF;
                ide.drive=(val>>4)&1;
//                rpclog("Write IDE head %02X %i,%i\n",val,ide.head,ide.drive);
                return;
                case 0x1F7: /*Command register*/
  //              rpclog("Starting new command %02X\n",val);
                ide.command=val;
                ide.error=0;
                switch (val)
                {
                        case 0x10: /*Restore*/
                        case 0x70: /*Seek*/
                        ide.atastat=0x40;
                        idecallback=100;
                        return;
                        case 0x20: /*Read sector*/
/*                        if (ide.secount>1)
                        {
                                error("Read %i sectors from sector %i cylinder %i head %i\n",ide.secount,ide.sector,ide.cylinder,ide.head);
                                exit(-1);
                        }*/
//                        rpclog("Read %i sectors from sector %i cylinder %i head %i\n",ide.secount,ide.sector,ide.cylinder,ide.head);
                        ide.atastat=0x80;
                        idecallback=100;
                        return;
                        case 0x30: /*Write sector*/
/*                        if (ide.secount>1)
                        {
                                error("Write %i sectors to sector %i cylinder %i head %i\n",ide.secount,ide.sector,ide.cylinder,ide.head);
                                exit(-1);
                        }*/
//                        rpclog("Write %i sectors to sector %i cylinder %i head %i\n",ide.secount,ide.sector,ide.cylinder,ide.head);
                        ide.atastat=0x08;
                        ide.pos=0;
                        return;
                        case 0x40: /*Read verify*/
                        case 0x41:
//                        rpclog("Read verify %i sectors from sector %i cylinder %i head %i\n",ide.secount,ide.sector,ide.cylinder,ide.head);
                        ide.atastat=0x80;
                        idecallback=200;
                        return;
                        case 0x50:
//                        rpclog("Format track %i head %i\n",ide.cylinder,ide.head);
                        ide.atastat=0x08;
//                        idecallback=200;
                        ide.pos=0;
                        return;
                        case 0x91: /*Set parameters*/
                        ide.atastat=0x80;
                        idecallback=200;
                        return;
                        case 0xA1: /*Identify packet device*/
                        case 0xE3: /*Idle*/
                        ide.atastat=0x80;
                        idecallback=200;
                        return;
                        case 0xEC: /*Identify device*/
                        ide.atastat=0x80;
                        idecallback=200;
                        return;
                        case 0xE5: /*Standby power check*/
                        ide.atastat=0x80;
                        idecallback=200;
                        return;
                }
                error("Bad IDE command %02X\n",val);
                exit(-1);
                return;
                case 0x3F6:
                if ((ide.fdisk&4) && !(val&4))
                {
                        idecallback=500;
                        idereset=1;
                        ide.atastat=0x80;
//                        rpclog("IDE Reset\n");
                }
                ide.fdisk=val;
                return;
        }
        error("Bad IDE write %04X %02X\n",addr,val);
        dumpregs();
        exit(-1);
}

unsigned char readide(unsigned short addr)
{
        unsigned char temp;
//        rpclog("Read IDE %08X %08X\n",addr,PC-8);
        switch (addr)
        {
                case 0x1F0:
//                rpclog("Read data %08X %02X\n",ide.pos,idebufferb[ide.pos]);
/*                if (!dumpedread)
                {
                        f=fopen("ram212.dmp","wb");
                        for (c=0x2127800;c<0x2127C00;c++)
                        {
                                putc(readmemb(c),f);
                        }
                        fclose(f);
                }*/
                temp=idebufferb[ide.pos];
                ide.pos++;
                if (ide.pos>=512)
                {
                        ide.pos=0;
                        ide.atastat=0x40;
                }
                return temp;
                case 0x1F1:
//                rpclog("Read IDEerror %02X\n",ide.atastat);
                return ide.error;
                case 0x1F2:
                return ide.secount;
                case 0x1F3:
                return ide.sector;
                case 0x1F4:
                return ide.cylinder&0xFF;
                case 0x1F5:
                return ide.cylinder>>8;
                case 0x1F6:
//                        rpclog("Read IDE Head Drive %02X %02X\n",ide.head|(ide.drive<<4),armregs[1]);
                return ide.head|(ide.drive<<4)|0xA0;
                case 0x1F7:
                if (romset==3)
                {
                        ioc.irqb&=~8;
                        updateirqs();
                }
//                rpclog("Read ATAstat %02X\n",ide.atastat);
                return ide.atastat;
                case 0x3F6:
//                rpclog("Read ATAstat %02X\n",ide.atastat);
                return ide.atastat;
        }
        error("Bad IDE read %04X\n",addr);
        dumpregs();
        exit(-1);
}

unsigned short readidew()
{
        unsigned short temp;
//        if (ide.sector==7) rpclog("Read data2 %08X %04X\n",ide.pos,idebuffer[ide.pos>>1]);
        temp=idebuffer[ide.pos>>1];
        ide.pos+=2;
        if (ide.pos>=512)
        {
                ide.pos=0;
                ide.atastat=0x40;
                if (ide.command==0x20)
                {
                        ide.secount--;
//                        rpclog("Sector done - secount %i\n",ide.secount);
                        if (ide.secount)
                        {
                                ide.atastat=0x08;
                                ide.sector++;
                                if (ide.sector==(ide.spt[ide.drive]+1))
                                {
                                        ide.sector=1;
                                        ide.head++;
                                        if (ide.head==(ide.hpc[ide.drive]))
                                        {
                                                ide.head=0;
                                                ide.cylinder++;
                                        }
                                }
                                ide.atastat=0x80;
                                idecallback=200;
                        }
                }
        }
        return temp;
}

void callbackide()
{
        int addr,c;
        if (idereset)
        {
                ide.atastat=0x40;
                ide.error=0;
                ide.secount=1;
                ide.sector=1;
                ide.head=0;
                ide.cylinder=0;
                idereset=0;
//                rpclog("Reset callback\n");
                return;
        }
        switch (ide.command)
        {
                case 0x10: /*Restore*/
                case 0x70: /*Seek*/
//                rpclog("Restore callback\n");
                ide.atastat=0x40;
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0x20: /*Read sectors*/
                if (!ide.secount)
                {
                        ide.atastat=0x40;
                        return;
                }
                readflash[0]=1;
                addr=((((ide.cylinder*ide.hpc[ide.drive])+ide.head)*ide.spt[ide.drive])+(ide.sector))*512;
                if (!skip512[ide.drive]) addr-=512;
                /*                if (ide.cylinder || ide.head)
                {
                        error("Read from other cylinder/head");
                        exit(-1);
                }*/
                fseek(hdfile[ide.drive],addr,SEEK_SET);
                fread(idebuffer,512,1,hdfile[ide.drive]);
                ide.pos=0;
                ide.atastat=0x08;
//                rpclog("Read sector callback %i %i %i offset %08X %i left %i\n",ide.sector,ide.cylinder,ide.head,addr,ide.secount,ide.spt[ide.drive]);
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0x30: /*Write sector*/
                readflash[0]=2;
                addr=((((ide.cylinder*ide.hpc[ide.drive])+ide.head)*ide.spt[ide.drive])+(ide.sector))*512;
                if (!skip512[ide.drive]) addr-=512;
//                rpclog("Write sector callback %i %i %i offset %08X %i left %i %i %i\n",ide.sector,ide.cylinder,ide.head,addr,ide.secount,ide.spt[ide.drive],ide.hpc[ide.drive],ide.drive);
                fseek(hdfile[ide.drive],addr,SEEK_SET);
                fwrite(idebuffer,512,1,hdfile[ide.drive]);
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                ide.secount--;
                if (ide.secount)
                {
                        ide.atastat=0x08;
                        ide.pos=0;
                        ide.sector++;
                        if (ide.sector==(ide.spt[ide.drive]+1))
                        {
                                ide.sector=1;
                                ide.head++;
                                if (ide.head==(ide.hpc[ide.drive]))
                                {
                                        ide.head=0;
                                        ide.cylinder++;
                                }
                        }
                }
                else
                   ide.atastat=0x40;
                return;
                case 0x40: /*Read verify*/
                case 0x41:
                ide.pos=0;
                ide.atastat=0x40;
//                rpclog("Read verify callback %i %i %i offset %08X %i left\n",ide.sector,ide.cylinder,ide.head,addr,ide.secount);
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0x50: /*Format track*/
                addr=(((ide.cylinder*ide.hpc[ide.drive])+ide.head)*ide.spt[ide.drive])*512;
                if (!skip512[ide.drive]) addr-=512;
//                rpclog("Format cyl %i head %i offset %08X secount %I\n",ide.cylinder,ide.head,addr,ide.secount);
                fseek(hdfile[ide.drive],addr,SEEK_SET);
                memset(idebufferb,0,512);
                for (c=0;c<ide.secount;c++)
                {
                        fwrite(idebuffer,512,1,hdfile[ide.drive]);
                }
                ide.atastat=0x40;
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0x91: /*Set parameters*/
                ide.spt[ide.drive]=ide.secount;
                ide.hpc[ide.drive]=ide.head+1;
//                rpclog("%i sectors per track, %i heads per cylinder  %i %i  %i\n",ide.spt[ide.drive],ide.hpc[ide.drive],ide.secount,ide.head,ide.drive);
                ide.atastat=0x40;
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0xA1:
                case 0xE3:
                        case 0xE5:
                ide.atastat=0x41;
                ide.error=4;
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
                case 0xEC:
//                        rpclog("Callback EC\n");
                memset(idebuffer,0,512);
                idebuffer[1]=101; /*Cylinders*/
                idebuffer[3]=16;  /*Heads*/
                idebuffer[6]=63;  /*Sectors*/
                for (addr=10;addr<20;addr++)
                    idebuffer[addr]=0x2020;
                for (addr=23;addr<47;addr++)
                    idebuffer[addr]=0x2020;
                idebufferb[46^1]='v'; /*Firmware version*/
                idebufferb[47^1]='0';
                idebufferb[48^1]='.';
                idebufferb[49^1]='5';
                idebufferb[54^1]='A'; /*Drive model*/
                idebufferb[55^1]='r';
                idebufferb[56^1]='c';
                idebufferb[57^1]='u';
                idebufferb[58^1]='l';
                idebufferb[59^1]='a';
                idebufferb[60^1]='t';
                idebufferb[61^1]='o';
                idebufferb[62^1]='r';
                idebufferb[63^1]='H';
                idebufferb[64^1]='D';
                idebuffer[50]=0x4000; /*Capabilities*/
                ide.pos=0;
                ide.atastat=0x08;
//                rpclog("ID callback\n");
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;
/*                case 0xE5:
                ide.
                ide.atastat=0x40;
                if (romset==3)
                {
                        ioc.irqb|=8;
                        updateirqs();
                }
                return;*/
        }
}
/*Read 1F1*/
/*Error &108A1 - parameters not recognised*/
